File: server/Services/UserService.cs
using Data.Entities;
using Data.Repositories;
using Dtos.User;
using Services;
using Utils;

namespace Services;

public class UserService : IUserService
{
  private readonly IUserRepository _userRepository;

  public UserService(IUserRepository userRepository)
  {
    _userRepository = userRepository;
  }

  public async Task<ServiceResponse<UsersEntity>> AddUser(UserPostDto user)
  {
    var serviceResponse = new ServiceResponse<UsersEntity>();

    try
    {
      var isUserExists = await _userRepository.UserExists(user.Uid);
      if (isUserExists)
      {
        serviceResponse.Data = null;
        serviceResponse.Success = false;
        serviceResponse.ErrorMessage = "User already exists with this email address";
        return serviceResponse;
      }

      var newUser = new UsersEntity()
      {
        Name = user.Name,
        Email = user.Email,
        PhotoUrl = user.PhotoUrl,
        Uid = user.Uid
      };
      var repoResponse = await _userRepository.AddUser(newUser);
      if (repoResponse.Success)
      {
        serviceResponse.Data = repoResponse.Data;
        serviceResponse.Success = true;
        serviceResponse.ErrorMessage = "";
        return serviceResponse;
      }

      serviceResponse.Success = false;
      return serviceResponse;
    }
    catch (Exception ex)
    {
      serviceResponse.Data = null;
      serviceResponse.ErrorMessage = ex.ToString();
      serviceResponse.Success = true;

      return serviceResponse;
    }
  }
}

File: server/src/services/UserService.ts
import UserRepository from '../database/repositories/UserRepository.js';
import IUserService from './interfaces/IUserService.js';
import UserPostDto from '../dtos/user/UserPostDto.js';
import ServiceResponse from '../utils/ServiceResponse.js';
import UserUpdateDto from 'src/dtos/user/UserUpdateDto.js';
import RepoResponse from 'src/utils/RepoResponse.js';

export class UserService implements IUserService {
  private async isUserFoundByUid(uid: string): Promise<boolean> {
    if (uid) return await UserRepository.isUserFound(uid);
    return false;
  }

  private async getUserIdByUid(uid: string): Promise<string | null> {
    if (uid) {
      // Check if the user exists
      const isUserFound = await this.isUserFoundByUid(uid);
      if (!isUserFound) {
        return null;
      }

      // User exists, then get user information
      const repoResponse = await UserRepository.findOneIdByUid(uid);
      if (!repoResponse.isSuccess) {
        return null;
      }

      return repoResponse.data.id;
    }
    return null;
  }

  public async createUser(userToUpdate: UserPostDto): Promise<ServiceResponse> {
    const serviceResponse = new ServiceResponse();

    try {
      // Check if user already exists
      const userId = await this.getUserIdByUid(userToUpdate.uid);
      console.log(userId);
      if (!userId) {
        // User does not exist, create them
        const createdUser = await UserRepository.insertOne(userToUpdate);
        return await serviceResponse.success(createdUser.data);
      }

      // User already exists, update their information
      const repoResponse = await UserRepository.updateOne(userId, userToUpdate);
      if (!repoResponse.isSuccess) {
        return await serviceResponse.createFromError(repoResponse.errorMessage);
      }

      return await serviceResponse.success(repoResponse.data);
    } catch (error: any) {
      console.error(error);
      return await serviceResponse.createFromError(error.toString());
    }
  }

  public async getUser(uid: string | null): Promise<ServiceResponse> {
    const serviceResponse = new ServiceResponse();
    try {
      if (!uid) {
        return await serviceResponse.createFromError(
          'Please enter a correct uid',
        );
      }

      // Get User Id
      const userId = await this.getUserIdByUid(uid);
      if (!userId) {
        return await serviceResponse.createFromError('User Not Found');
      }

      const user = await UserRepository.findOne(userId);
      return await serviceResponse.success(user.data);
    } catch (error: any) {
      return await serviceResponse.createFromError(error.toString());
    }
  }

  public async updateUser(
    uid: string,
    user: UserUpdateDto,
  ): Promise<ServiceResponse> {
    const serviceResponse = new ServiceResponse();
    try {
      if (!uid) {
        return await serviceResponse.createFromError(
          'Please enter a correct uid',
        );
      }

      const userId = await this.getUserIdByUid(uid);
      console.log(userId);
      if (!userId) {
        return await serviceResponse.createFromError('User Not Found');
      }

      const updatedUser: RepoResponse = await UserRepository.updateOne(
        userId,
        user,
      );
      if (!updatedUser.isSuccess) {
        return await serviceResponse.createFromError(updatedUser.errorMessage);
      }

      return await serviceResponse.success(updatedUser.data);
    } catch (error: any) {
      console.log(error);
      return await serviceResponse.createFromError(error.toString());
    }
  }

  async deleteUser(uid: string): Promise<ServiceResponse> {
    throw new Error('Method not implemented.');
  }
}

export default new UserService();
